src/main/scala/com/microsoft/hyperspace/index/rules/RuleUtils.scala

```diff
@@ -106,4 +128,318 @@ object RuleUtils {
       None // logicalPlan is non-linear or it has no LogicalRelation.
     }
   }
+
+  /**
+   * Transform the current plan to utilize the given index.
+   *
+   * The transformed plan reads the given index data rather than original source files.
+   * If HybridScan is enabled, additional logical plans for the appended source files would be
+   * constructed and merged with the transformed plan with the index.
+   *
+   * Pre-requisites
+   * - We know for sure the index which can be used to replace the plan.
+   * - The plan should be linear and include 1 LogicalRelation.
+   *
+   * @param spark Spark session.
+   * @param index Index used in replacement plan.
+   * @param plan Current logical plan.
+   * @param useBucketSpec Option whether to use BucketSpec for reading index data.
+   * @return Replacement plan.
+   */
+  def transformPlanToUseIndex(
+      spark: SparkSession,
+      index: IndexLogEntry,
+      plan: LogicalPlan,
+      useBucketSpec: Boolean): LogicalPlan = {
+    val transformed = if (HyperspaceConf.hybridScanEnabled(spark)) {
+      transformPlanToUseHybridIndexDataScan(spark, index, plan, useBucketSpec)
+    } else {
+      transformPlanToUsePureIndexScan(spark, index, plan, useBucketSpec)
+    }
+    assert(!transformed.equals(plan))
```

Comment:
I think this `assert` should be replaced by an `exception` thrown.
Imagine the replacement fails and no transformation happens. I think, This assert basically makes the whole user query to break. One functionality aspect of Hyeprspace is that: "A user's query should not really break due to the "extra" optimization Hyperspace tries to add". That is why in the rules, we used `Exceptions` and in the `Catch` block of `transofrm` we catch any exception, log it and return the original plan. We dont really wanna fail a query if our attempt to (further) optimize fails.

Reply:
It's because there are pre-requite conditions and we previously check conditions for this like linear plan, candidate indexes, .. etc. I think some asserts are required for development & testing. If we only throw exceptions which are ignored in Catch block, it's hard to debug any issue or even unnoticed.
This assert was suggested by @imback82 in #165.

Reply:
Yes, assert makes more sense than exception here since we are validating the "assumption" for this function.